acceptance: TestDockerCLI_test_demo_multitenant failed

acceptance.TestDockerCLI_test_demo_multitenant [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_Acceptance/18919015?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_Acceptance/18919015?buildTab=artifacts#/) on master @ [a2882ae5da176e086393753f11e7daf6bf68ff8a](https://github.com/cockroachdb/cockroach/commits/a2882ae5da176e086393753f11e7daf6bf68ff8a):


```
=== RUN   TestDockerCLI_test_demo_multitenant
    test_log_scope.go:165: test logs captured to: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/_tmp/f14cb1cd56830b137df13910221b8384/logTestDockerCLI_test_demo_multitenant3440821787
    test_log_scope.go:76: use -show-logs to present logs inline
    generated_cli_test.go:126: -- test log scope end --
test logs left over in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance/_tmp/f14cb1cd56830b137df13910221b8384/logTestDockerCLI_test_demo_multitenant3440821787
--- FAIL: TestDockerCLI_test_demo_multitenant (72.05s)
=== RUN   TestDockerCLI_test_demo_multitenant/test_demo_multitenant.tcl
    util_docker.go:256: non-zero exit code: 1
    --- FAIL: TestDockerCLI_test_demo_multitenant/test_demo_multitenant.tcl (56.89s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDockerCLI_test_demo_multitenant.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/acceptance/generated_cli_test.go
+++ b/pkg/acceptance/generated_cli_test.go
@@ -118,6 +118,13 @@ func TestDockerCLI_test_demo_memory_warning(t *testing.T) {
 	runTestDockerCLI(t, "test_demo_memory_warning", "../cli/interactive_tests/test_demo_memory_warning.tcl")
 }
 
+func TestDockerCLI_test_demo_multitenant(t *testing.T) {
+	s := log.Scope(t)
+	defer s.Close(t)
+
+	runTestDockerCLI(t, "test_demo_multitenant", "../cli/interactive_tests/test_demo_multitenant.tcl")
+}
+
 func TestDockerCLI_test_demo_networking(t *testing.T) {
 	s := log.Scope(t)
 	defer s.Close(t)

--- a/pkg/cli/clisqlshell/sql.go
+++ b/pkg/cli/clisqlshell/sql.go
@@ -66,6 +66,7 @@ Connection
                     connect to a server or print the current connection URL.
                     Omitted values reuse previous parameters. Use '-' to skip a field.
                     The option "autocerts" attempts to auto-discover TLS client certs.
+                    The option "tenant=NAME" switches to the specified tenant.
   \password [USERNAME]
                     securely change the password for a user
 
@@ -1708,6 +1709,11 @@ func (c *cliState) handleConnectInternal(cmd []string, omitConnString bool) erro
 		if cmd[4] != "-" {
 			if cmd[4] == "autocerts" {
 				autoCerts = true
+			} else if strings.HasPrefix(cmd[4], "tenant=") {
+				tenantName := strings.TrimPrefix(cmd[4], "tenant=")
+				if err := newURL.SetOption("options", "-ccluster="+tenantName); err != nil {
+					return err
+				}
 			} else {
 				return errors.Newf(`unknown syntax: \c %s`, strings.Join(cmd, " "))
 			}

--- a/pkg/cli/interactive_tests/test_demo_multitenant.tcl
+++ b/pkg/cli/interactive_tests/test_demo_multitenant.tcl
@@ -1,10 +1,10 @@
 #! /usr/bin/env expect -f
 
-# TODO(141942): Re-enable this test once we have investigated its flakiness.
-
 source [file join [file dirname $argv0] common.tcl]
 
-spawn $argv demo --no-line-editor --empty --nodes 3 --multitenant --log-dir=logs
+set ::env(COCKROACH_DEMO_PASSWORD) "hunter2hunter2hunter2hunter2"
+
+spawn $argv demo --no-line-editor --empty --nodes 5 --multitenant --log-dir=logs
 
 start_test "Check that the demo cluster starts properly"
 eexpect "Welcome"
@@ -43,31 +43,67 @@ eexpect "defaultdb>"
 end_test
 
 start_test "Check that the demo add command adds a node with no error in ls"
-send "\\demo add region=us-east-1\r"
-eexpect "node 4 has been added"
+send "\\demo add region=us-west1\r"
+eexpect "node 6 has been added"
 eexpect "defaultdb>"
 send "\\demo ls\r"
 eexpect "node 1:"
 eexpect "node 2:"
 eexpect "node 3:"
 eexpect "node 4:"
+eexpect "node 5:"
+eexpect "node 6:"
 eexpect "Application tenant:"
 eexpect "System tenant:"
 eexpect "defaultdb>"
 end_test
 
 start_test "Check that a node can be restarted"
-send "\\demo shutdown 4\r"
-eexpect "node 4 has been shutdown"
+
+# Switch to system tenant to check the liveness range.
+send "\\connect - - - - tenant=system\r"
+eexpect "the cluster ID has changed!"
 eexpect "defaultdb>"
-send "\\demo restart 4\r"
-eexpect "node 4 has been restarted"
+
+# Wait for the liveness range to have the default 5 voters. If its replication
+# factor is too low, shutting down the node below can cause it to lose quorum
+# and stall the shutdown command (example: #147867).
+set timeout 2
+set stmt "select range_id, array_length(voting_replicas,1) from crdb_internal.ranges where range_id=2;\r"
+send $stmt
+expect {
+    "2 |            5" {
+        puts "\rliveness range has 5 voters"
+    }
+    timeout {
+        puts "\rliveness range does not yet have 5 voters"
+        sleep 2
+        send $stmt
+        exp_continue
+    }
+}
+# Reset timeout back to 45 to match common.tcl.
+set timeout 45
+eexpect "defaultdb>"
+
+# Switch back to application tenant for the rest of the test.
+send "\\connect - - - - tenant=demoapp\r"
+eexpect "the cluster ID has changed!"
+eexpect "defaultdb>"
+
+send "\\demo shutdown 6\r"
+eexpect "node 6 has been shutdown"
+eexpect "defaultdb>"
+send "\\demo restart 6\r"
+eexpect "node 6 has been restarted"
 eexpect "defaultdb>"
 send "\\demo ls\r"
 eexpect "node 1:"
 eexpect "node 2:"
 eexpect "node 3:"
 eexpect "node 4:"
+eexpect "node 5:"
+eexpect "node 6:"
 eexpect "Application tenant:"
 eexpect "System tenant:"
 eexpect "defaultdb>"

